nautobot/core/app/__init__.py

```diff
@@ -0,0 +1,244 @@
+from django.apps import AppConfig
+from django.urls import reverse
+from operator import getitem
+from collections import OrderedDict
+
+from nautobot.extras.plugins.utils import import_object
+from nautobot.extras.registry import registry
+from nautobot.utilities.choices import ButtonActionColorChoice
+
+
+registry["nav_menu"] = {"tabs": {}}
+
+
+class NautobotConfig(AppConfig):
+    """
+    Custom AppConfig for Nautobot application.
+
+    Adds functionality to generate the HTML navigation menu using `navigation.py` files from nautbot
+    applications.
+    """
+
+    menu_tabs = "navigation.menu_tabs"
+
+    def ready(self):
+        """
+        Ready function initiates the import application.
+
+        As the application can be loaded from either the root of the app or `apps.py`, the __module__
+        name needs to be modified to all the ready function to import the menu items.
+
+        To import the menu items the `.apps` needs to be removed from the __module__ name if it
+        exists.
+
+        Examples:
+            Loaded from application root:   `nautobot.circuits`
+            Loaded from `app.py`:           `nautobot.circuits.apps`
+        """
+        if self.__module__.endswith(".apps"):
+            self.__module__ = self.__module__[:-5]
+        menu_items = import_object(f"{self.__module__}.{self.menu_tabs}")
+        if menu_items is not None:
+            register_menu_items(menu_items)
+
+
+def register_menu_items(tab_list):
+    """
+    Using the imported object a dictionary is either created or updated with objects to create
+    the navbar.
+
+    The dictionary is built from four key objects, NavMenuTab, NavMenuGroup, NavMenuItem and
+    NavMenuButton. The Django template then uses this dictionary to generate the navbar HTML.
+    """
+    for nav_tab in tab_list:
+        if nav_tab.name not in registry["nav_menu"]["tabs"]:
+            registry["nav_menu"]["tabs"][nav_tab.name] = {
+                "weight": nav_tab.weight,
+                "groups": registry["nav_menu"]["tabs"].get(nav_tab.name, {}).get("groups", {}),
```

Comment:
Since the `if nav_tab.name not in registry...` is added above, you're always adding a new entry here and so you don't need to check for existing data under `nav_tab.name`.

```suggestion
                "groups": {},
```